src/models/d2ModelFactory.js

```diff
@@ -0,0 +1,24 @@
+import {
+    DataElementModel,
+    defaultModel,
+    IndicatorModel,
+    OrganisationUnitModel,
+    ValidationRuleModel,
+} from "./d2Model";
+
+const classes = {
+    OrganisationUnitModel,
+    DataElementModel,
+    IndicatorModel,
+    ValidationRuleModel,
+};
+
+/**
+ * D2ModelProxy allows to create on-demand d2Model classes
+ * If the class doesn't exist a new default class is created
+ * d2ModelName: string (singular name property from d2.models)
+ */
+export function d2ModelFactory(d2ModelName) {
+    const modelName = d2ModelName.charAt(0).toUpperCase() + d2ModelName.slice(1) + "Model";
+    return classes[modelName] ? classes[modelName] : defaultModel(d2ModelName);
+}
```

Comment:
TS shines when using those model classes, I am a bit reluctant to leave it as JS, callers won't take advantage of typings/autocomplete. If we could have fixed classes, this should work:

```
type Classes = typeof classes;

export function d2ModelFactory<K extends keyof Classes>(modelName: K): Classes[K] {
    return classes[modelName];
}
```

If having the ability to create classes on-the-fly (as you have know) is a must, this is not so type-restrictive, but  should also work:

```
const classes: { [modelName: string]: D2Model } = {
    OrganisationUnitModel,
    DataElementModel,
    IndicatorModel,
    ValidationRuleModel,
};

export function d2ModelFactory(d2ModelName: string): D2Model {
    const modelName = d2ModelName.charAt(0).toUpperCase() + d2ModelName.slice(1) + "Model";
    return classes[modelName] ? classes[modelName] : defaultModel(d2ModelName);
}
```

Reply:
Thanks!

Reply:
Actually we do not return a D2Model instance but the class of D2Model.

```js
function d2ModelFactory(d2ModelName: string): typeof D2Model
```